CoreWordCloud/src/au/gov/asd/tac/constellation/views/wordcloud/content/ContentVectorClusteringServices.java

```diff
@@ -0,0 +1,316 @@
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.concurrent.ConcurrentSkipListMap;
+
+/**
+ *
+ * @author twilight_sparkle
+ */
+public class ContentVectorClusteringServices {
+
+    private VectorWeightingCalculator weightingCalculator;
+    private Clustering clusteringMethod;
+    private final SparseMatrix<Integer> tokenElementMatrix;
+    private final SparseMatrix<Float> elementTokenMatrix;
+    private final Map<Integer, Integer> moduli;
+    private final Map<Integer, Integer> elementToCluster;
+    private final Map<Integer, List<Integer>> clusters;
+    private final List<Integer> clusterCentres;
+    private static final Random r = new Random();
+
+    private ContentVectorClusteringServices(final SparseMatrix<Integer> tokenElementMatrix) {
+        this.tokenElementMatrix = tokenElementMatrix;
+        this.elementTokenMatrix = SparseMatrix.constructMatrix((Float) 0F);
+        this.elementToCluster = new ConcurrentSkipListMap<>();
+        this.clusterCentres = new ArrayList<>();
+        moduli = new ConcurrentSkipListMap<>();
+        clusters = new ConcurrentSkipListMap<>();
+    }
+
+    public enum WeightingMethod {
+        APPEARANCE,
+        RANK;
+    }
+
+    public static ContentVectorClusteringServices createKMeansClusteringService(final DefaultTokenHandler handler, final ClusterDocumentsParameters clusterDocumentsParams, final int numberOfElements) {
+        final ContentVectorClusteringServices cvcs = new ContentVectorClusteringServices(handler.getTokenElementMatrix());
+        cvcs.clusteringMethod = cvcs.new KMeansClustering(clusterDocumentsParams.getNumberOfMeans());
+        if (clusterDocumentsParams.getThresholdMethod().equals(TokenThresholdMethod.APPEARANCE)) {
+            cvcs.weightingCalculator = cvcs.new CountAppearancesCalculator(numberOfElements, clusterDocumentsParams.getThresholdPercentage(), true, clusterDocumentsParams.isBinarySpace(), clusterDocumentsParams.isSignificantAboveThreshold(), clusterDocumentsParams.getWeightingExponent());
+        } else {
+            cvcs.weightingCalculator = cvcs.new RankTokenCalculator(numberOfElements, cvcs.tokenElementMatrix.getNumColumns(), clusterDocumentsParams.getThresholdPercentage(), true, clusterDocumentsParams.isBinarySpace(), clusterDocumentsParams.isSignificantAboveThreshold(), clusterDocumentsParams.getWeightingExponent());
+        }
+        return cvcs;
+    }
+    
+    protected VectorWeightingCalculator getWeightingCalculator() {
+        return weightingCalculator;
+    }
+    
+    protected SparseMatrix<Integer> getTokenElementMatrix() {
+        return tokenElementMatrix;
+    }
+
+    private interface Clustering {
+
+        public void cluster();
+    }
+
+    public Map<Integer, Integer> getClusters() {
+        return new HashMap<>(elementToCluster);
```

Comment:
does this need a new map? If trying to avoid editing existing one, `Collections.unmodifiableMap` would be better